lsp: Fix workspace diagnostics lag & add streaming support

4teapo opened the pull request:
Closes https://github.com/zed-industries/zed/issues/33980
Closes https://github.com/zed-industries/zed/discussions/33979

- Switches to the debounce task pattern for diagnostic summary computations, which most importantly lets us do them only once when a large number of DiagnosticUpdated events are received at once.
- Makes workspace diagnostic requests not time out if a partial result is received.
- Makes diagnostics from workspace diagnostic partial results get merged. 

There might be some related areas where we're not fully complying with the LSP spec but they may be outside the scope of what this PR should include.

Release Notes:

- Added support for streaming LSP workspace diagnostics.
- Fixed editor freeze from large LSP workspace diagnostic responses.

4teapo committed:
lsp: Fix workspace diagnostics lag & add streaming support

- Adds DiagnosticsBatchUpdated events, emitted whenever some diagnostics
batch (1 or more diagnostics) is processed. These are always emitted
alongside DiagnosticsUpdated evenst, but only once for each group of
diagnostics. - Moves some things from DiagnosticsUpdated to
DiagnosticsBatchUpdated for performance improvement. - Removes timeout
for workspace diagnostics request, so that they can be long-running for
streaming purposes, & adds partial results handling for it.

There might be some areas where we're not fully complying with the LSP
spec but they may be outside the scope of this.

4teapo committed:
Merge branch 'zed-industries:main' into partial-workspace-diagnostic

4teapo committed:
Add DiagnosticsBatchUpdated assert to test

4teapo committed:
Remove lsp from project_panel dependencies

4teapo committed:
Fix test not expecting DiagnosticsBatchUpdated

SomeoneToIgnore reviewed (commented):
Thank you a lot for coming up with the investigation and the PR: the forever-hanging request was sort of a surprise to me, and I wonder, could we test it and handle a bit better, if my fantasies about the spec and partial responses make sense?

Also, let's take care of the types better and see if we can bump Zed's lsp types library instead of introducing a custom handler.

Secondly, it would be great to rework the new event approach a bit: either by postponing any git diff calculations with a debounce, in a task (those events do not need to be waited upon synchronously) or at least, by uniting the new event's idea into the previous one, so we explicitly take care of all event-related, existing places.

Overall, seems so good that the test might be the hardest part that's left.

SomeoneToIgnore commented on crates/project/src/lsp_store.rs:
```diff
@@ -11434,72 +11534,11 @@ fn lsp_workspace_diagnostics_refresh(
-                                        }
-                                    }
-                                }
+                                lsp_store.apply_workspace_diagnostic_report(
```
I am a bit confused still as see no explicit description on what server can and can not do with both ways of sending the data back (via `$/progress` and via the workspace diagnostics pull responses).

I fully agree with the normal path, routed via `$/progress` and fine to keep it the only one covered if you think so.
What we miss here is a thing that makes it obvious that the distant bits connect: that's why a hope that we can write a test.

We could also move all functions/methods related to the disagnostics pull into a submodule within `lsp_store`, so it's more visible that the code is related.

4teapo committed:
Use simple workspace diagnostic partial result token

SomeoneToIgnore commented on crates/diagnostics/src/items.rs:
```diff
@@ -134,7 +134,7 @@ impl DiagnosticIndicator {
             }
 
-            project::Event::DiagnosticsUpdated { .. } => {
+            project::Event::DiagnosticsBatchUpdated => {
```
I think it's a totally different direction.
What I want to fully avoid, is an extra state management, batch management and other extra knowledge related.

If we take a step back, our issue seems to be in the fact, that certain global entities like project panel and diagnostics indicator react on each file diagnostics update synchronously.
We send diagnostics updates per file, and might send a lot of updates at once now, causing those synchronous reactions to call something heavy enough to block things.

When looking at all `DiagnosticsUpdated` exact word usages, it seems surprisingly small:

![image](https://github.com/user-attachments/assets/85dd2611-76ee-483d-9f5b-b1b6739f6b4e)

It's basically buffer synch mechanisms that bubble up, either caused by the buffer state sync mechanisms, or the diagnostics update - each happen per buffer.

What can we do, without introducing any extra "clever" state, to batch this?

1. We always know whether it's one file (due to diagnostics push or buffer state sync) or multiple, so can, in theory, pass some `send_event: bool` extra parameter somewhere on the `update_buffer_diagnostics` in the `LspStore`, and then send the same kind of "diagnostics updated" event from the lsp store level, after all diagnostics are applied.

We would need to change the project/lsp_store events to contain a `Vec<_>` of values, and introduce ~medium change to the codebase around the buffer sync.
Due to background remote buffer sync, the remote part will receive a lot of update messages anyway, but those will be passing through the wire, in a slower pace, causing less blocking.

2. We can do less work on such "diagnostics updated" events.

Due to the nature of our batches (born out of pull diagnostics), we do not get any frequent updates per single buffer, it's all different buffers.

So, reacting on such events in editors is still fine, and what's left are ~3 places where we recalculate things.
Some of these recaltulations happen per file path and seem to be ok to do, and pretty fast: it's the diagnostic summaries that usually cause issues it seems, and whatever else that got moved away into `DiagnosticsBatchUpdated` event handling branches.

Given that it's so few places that cause actual issues, we can do things fully asynchronously.
Inline diagnostics in the editor is a good example:

https://github.com/zed-industries/zed/blob/7432e947bc60204bc0f8fcfb13ab06f190369451/crates/editor/src/editor.rs#L16504-L16549

this code is called on any "diagnostics updated"-like call, fully traversing all diagnostics per file, and seem to work ok performance-wise.

This is achieved with a `inline_diagnostics_update: Task<()>` field in the `Editor`.
That field is initialized with `Task::ready(())` in the beginning and later, on each subsequent call, it creates another task that waits for a debounce (my 30ms proposed), then does all stateless, heavy operations in the `background_spawn`, off the main thread; regular state updates are available in the async task due to `cx.spawn_in(window, async move |editor, cx|`, outside the background spawn.

Using these primitives, we can always postpone the diagnostics summary calculation in the button and the project panel, and whatever remote or local source causing a flood of events, it is working ok, if we have a good debounce value?

-------------------------------

Neither options are great, but seem relatively small and bring no extra clever state and logic, really (that 2nd case is more like a pattern).

I lean towards the 2nd approach more, but happy with anything else that's matching the criteria.

4teapo committed:
Merge branch 'zed-industries:main' into partial-workspace-diagnostic

4teapo committed:
Merge branch 'partial-workspace-diagnostic' of https://github.com/4teapo/zed into partial-workspace-diagnostic

4teapo committed:
Merge main into partial-workspace-diagnostic

4teapo committed:
Add workspace diagnostic streaming tests and defer diagnostic summary
updates

We no longer do batching to determine when to update diagnostic
summaries, and instead make a task on individual DiagnosticsUpdated
events that updates the summary after 30 ms.

LSP requests can now have custom timers, and with this infrastructure
the `workspace/diagnostic` request gets timed out like other requests if
the response doesn't come within a certain time frame, unless a partial
result is received.

We also get rid of the LSP type overrides because the lsp-types crate
got updated to include them.

4teapo committed:
Merge remote-tracking branch 'origin/main' into partial-workspace-diagnostic

4teapo committed:
Merge branch 'main' into partial-workspace-diagnostic

4teapo committed:
Make expected_messages immutable

4teapo reviewed (commented):
Thanks for the great review! I've added a bit to editor tests, switched to the debounce task pattern, and some more, and I'm ready for round 2.

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -1126,13 +1127,36 @@ impl LanguageServer {
+    /// ready causes the request to be timed out with the future's output message.
+    ///
+    /// [LSP Specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage)
+    pub fn request_with_timer<T: request::Request, U: Future<Output = String> + Unpin>(
```
I think we're supposed to do more similar request in the future, e.g. there are workspace symbols request that might stream.

So, instead, we should keep it as is, ready for more timers to come?

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -1126,13 +1127,36 @@ impl LanguageServer {
+        )
+    }
+
+    fn request_internal_with_timer<T, U: Future<Output = String> + Unpin>(
```
Here in all methods: we have `where`, let's not pollute the `<T, U>` block with any bounds, and move all into `where` clause.

SomeoneToIgnore commented on crates/project/src/lsp_store.rs:
```diff
@@ -11379,7 +11464,7 @@ fn subscribe_to_binary_statuses(
     server: Arc<LanguageServer>,
     cx: &mut Context<'_, LspStore>,
-) -> Option<(mpsc::Sender<()>, Task<()>)> {
+) -> Option<(mpsc::Sender<()>, mpsc::Sender<()>, Task<()>)> {
```
At this point, should we introduce a struct with named fields for all this?
It is not very clear what each channel is doing by looking at this.

SomeoneToIgnore commented on crates/project/src/lsp_store.rs:
```diff
@@ -9130,67 +9133,77 @@ impl LspStore {
-                return;
-            };
+        match progress.value {
+            lsp::ProgressParamsValue::WorkDone(progress) => {
```
NIT: would be nice to extract this into another method too.

SomeoneToIgnore reviewed (commented):
This looks really nice and not invasive at all.
I also see no blocks in the `NevermoreEngine` project.

Thank you for putting it up together: I've left some minor notes, but we're almost ready to merge overall.

4teapo committed:
Improve style

Improves consistency prototypes for request with_timer methods, splits
WorkDoneProgress handling into its own method, moves main
`workspace_refresh_task` data to a struct.

SomeoneToIgnore reviewed (approved):
Thank you a lot for taking care of all this, that was very revealing spec-wise.
